### DISCUSSION TITLE
How to get a raw pointer of `vk::Device`?

### DISCUSSION THREAD

**Discussion by iddm** [ash-rs/ash#766] [Q&A]
I have a `ash::vk::Device` object, and I need to get a real vulkan `*mut vk::Device`. I see only the `Handle` trait, but the values there return an `u64`.

[I can only think of this hack:](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021)

```rust
fn main() {
    let number = 5u64;
    let pointer = std::ptr::null_mut::<u8>();
    let pointer = unsafe { pointer.add(number as _) };
    let vk_device_ptr: *mut ash::vk::Device = pointer.cast();
}
```

Regardless of the answer, can `ash` include such a possibility of returning native typed pointers to the user instead of just integers?

**Comment by MarijnS95** [comment]
`PFN_vkCreateDevice` also takes a `*mut vk::Device`, this already works by taking a borrow (which implicitly derefs to a raw pointer) to the `#[repr(transparent)]` `vk::Device` which is a `*mut u8` under the hood.

https://github.com/ash-rs/ash/blob/a40ddee3e271dd6f8b9249711864054ee126d8e7/ash/src/instance.rs#L364

https://github.com/ash-rs/ash/blob/a40ddee3e271dd6f8b9249711864054ee126d8e7/ash/src/vk/features.rs#L187


But the example/hack contradicts your request. What do you need this for, can you show an example of the API in question?

> and I need to get a real vulkan `*mut vk::Device`

What is meant here by "real"? Handles are typically passed _by value_ (except for getter cases like `vkCreateDevice()` demonstrated above) and are **pointer** `typedef`s to declared (but undefined!) `struct`s: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_DEFINE_HANDLE.html

**Reply by iddm** [reply]
That means that I had bindgen bindings to a c library which accepted a `vk::Device*`, so I couldn't work with just some transparent handle because of the types. I had to actually convert it to a pointer of the target type. So, thanks to the types being typedef as a pointer, I was able to calculate the offset myself. the way I did above (still not sure it is correct, though, but everything seems to work so far). So, it would be nice to have a trait and implement it for all these types to get a raw pointer to these objects, without doing what I did every single time.

**Reply by MarijnS95** [reply]
There is no API to add here, afaik. If you have an `ash::vk::Device`, you just take a borrow of that and let the compiler coerce that to a pointer, so that it becomes `*mut ash::vk::Device` which is equivalent to _a pointer to a `Device` handle_?

Can you share the library, header, and/or bindgen result so that we can take a look at what you need? Can you share what bindgen is `typedef`'ing that `vk::Device` to?
